### ISSUE TITLE
use view on variable cause backward fail to assign grad?

### ISSUE THREAD

**Issue by Frefreak** [pytorch/pytorch#1479]
Hi, I notice that if I use `view` on a variable, the variable's `grad` would always be `None`. I'm new to this library and couldn't find any doc mentions this? Is this the expected behavior?
```python
#!/usr/bin/env python

import torch
from torch.autograd import Variable

a = Variable(torch.Tensor([1, 2, 3])).view(1, 3)
b = Variable(torch.Tensor([4, 5, 6]), requires_grad=True).view(3, 1) # view applied to variable makes b.grad None
c = a.mm(b).sum()
c.backward()
print(a.requires_grad)
print(b.requires_grad)
print(b.grad)

print('-' * 20)

del a, b, c

a = Variable(torch.Tensor([1, 2, 3])).view(1, 3)
b = Variable(torch.Tensor([4, 5, 6]).view(3, 1), requires_grad=True) # view applied to tensor works
c = a.mm(b).sum()
c.backward()
print(a.requires_grad)
print(b.requires_grad)
print(b.grad)
```

output:
```
False
True
None
--------------------
False
True
Variable containing:
 1
 2
 3
[torch.FloatTensor of size 3x1]
```
I installed pytorch via `pip install http://download.pytorch.org/whl/cu80/torch-0.1.12.post2-cp36-cp36m-linux_x86_64.whl ` (version 0.1.12_2)

**Comment by soumith** [comment]
it's not a question of view vs non-view, we only populate gradients for leaf variables, i.e. Variables that were user-created.

```python
a = Variable(torch.Tensor([1, 2, 3])).view(1, 3) 
# not user-created, as it is seen as:
# _ = Variable(torch.Tensor([1, 2, 3]))
# a = _.view(1, 3)


b = Variable(torch.Tensor([4, 5, 6]), requires_grad=True).view(3, 1)
# still not user created. Variable is created and view is applied for it

b = Variable(torch.Tensor([4, 5, 6]).view(3, 1), requires_grad=True)
# user created. no operation is applied to the created Variable since it's created.
```

In the future feel free to search and ask questions on https://discuss.pytorch.org , there are a few threads discussing this specific topic.

**Comment by Frefreak** [comment]
Oh, I understand now... Thank you.
